\hypertarget{alignment_8h}{}\subsection{alignment.\+h File Reference}
\label{alignment_8h}\index{alignment.\+h@{alignment.\+h}}


File handling functions and calculation of distances for sequence data in nexus format.  


{\ttfamily \#include \char`\"{}distance\+\_\+matrix.\+h\char`\"{}}\newline
{\ttfamily \#include \char`\"{}nexus\+\_\+common.\+h\char`\"{}}\newline
Include dependency graph for alignment.\+h\+:\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=350pt]{alignment_8h__incl}
\end{center}
\end{figure}
This graph shows which files directly or indirectly include this file\+:\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=148pt]{alignment_8h__dep__incl}
\end{center}
\end{figure}
\subsubsection*{Data Structures}
\begin{DoxyCompactItemize}
\item 
struct \hyperlink{structalignment__struct}{alignment\+\_\+struct}
\begin{DoxyCompactList}\small\item\em Data from alignment file. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsubsection*{Typedefs}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{alignment_8h_ab92d2b81028457d293c6e56b2c2b815c}\label{alignment_8h_ab92d2b81028457d293c6e56b2c2b815c}} 
typedef struct \hyperlink{structalignment__struct}{alignment\+\_\+struct} $\ast$ {\bfseries alignment}
\end{DoxyCompactItemize}
\subsubsection*{Functions}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{alignment_8h_aa036d9585db0e7f896efff4e37ffd5cb}\label{alignment_8h_aa036d9585db0e7f896efff4e37ffd5cb}} 
\hyperlink{structalignment__struct}{alignment} \hyperlink{alignment_8h_aa036d9585db0e7f896efff4e37ffd5cb}{read\+\_\+alignment\+\_\+from\+\_\+file} (char $\ast$seqfilename)
\begin{DoxyCompactList}\small\item\em Reads D\+NA alignment (guess format between F\+A\+S\+TA and N\+E\+X\+US) from file and store info in \hyperlink{structalignment__struct}{alignment\+\_\+struct}. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_a57980b340fee11061ad4511efc7b7c47}\label{alignment_8h_a57980b340fee11061ad4511efc7b7c47}} 
\hyperlink{structalignment__struct}{alignment} \hyperlink{alignment_8h_a57980b340fee11061ad4511efc7b7c47}{read\+\_\+fasta\+\_\+alignment\+\_\+from\+\_\+file} (char $\ast$seqfilename)
\begin{DoxyCompactList}\small\item\em Reads D\+NA F\+A\+S\+TA alignment from file and store info in \hyperlink{structalignment__struct}{alignment\+\_\+struct}. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_a5bd5335c2fd2fcf8c30e1becfea64b2c}\label{alignment_8h_a5bd5335c2fd2fcf8c30e1becfea64b2c}} 
\hyperlink{structalignment__struct}{alignment} \hyperlink{alignment_8h_a5bd5335c2fd2fcf8c30e1becfea64b2c}{read\+\_\+nexus\+\_\+alignment\+\_\+from\+\_\+file} (char $\ast$seqfilename)
\begin{DoxyCompactList}\small\item\em Reads D\+NA N\+E\+X\+US alignment from file and store info in \hyperlink{structalignment__struct}{alignment\+\_\+struct}. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_ae7be6382b75167950cff01c19dbbe166}\label{alignment_8h_ae7be6382b75167950cff01c19dbbe166}} 
void \hyperlink{alignment_8h_ae7be6382b75167950cff01c19dbbe166}{print\+\_\+alignment\+\_\+in\+\_\+fasta\+\_\+format} (\hyperlink{structalignment__struct}{alignment} align, F\+I\+LE $\ast$stream)
\begin{DoxyCompactList}\small\item\em Prints alignment to F\+I\+LE stream in F\+A\+S\+TA format (debug purposes). \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_a98e39e61ac259230b72c60d7f033bdd4}\label{alignment_8h_a98e39e61ac259230b72c60d7f033bdd4}} 
void \hyperlink{alignment_8h_a98e39e61ac259230b72c60d7f033bdd4}{del\+\_\+alignment} (\hyperlink{structalignment__struct}{alignment} align)
\begin{DoxyCompactList}\small\item\em Frees memory from \hyperlink{structalignment__struct}{alignment\+\_\+struct}. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_a9f784a5ee1238ede5538c085b40cfbca}\label{alignment_8h_a9f784a5ee1238ede5538c085b40cfbca}} 
\hyperlink{structdistance__matrix__struct}{distance\+\_\+matrix} \hyperlink{alignment_8h_a9f784a5ee1238ede5538c085b40cfbca}{new\+\_\+distance\+\_\+matrix\+\_\+from\+\_\+valid\+\_\+matrix\+\_\+elems} (\hyperlink{structdistance__matrix__struct}{distance\+\_\+matrix} original, int $\ast$valid, int n\+\_\+valid)
\begin{DoxyCompactList}\small\item\em new matrix of pairwise distance by simply excluding original elements not present in valid\mbox{[}\mbox{]} \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_a51ff9ec227fbe812471f316a55bea52b}\label{alignment_8h_a51ff9ec227fbe812471f316a55bea52b}} 
\hyperlink{structdistance__matrix__struct}{distance\+\_\+matrix} \hyperlink{alignment_8h_a51ff9ec227fbe812471f316a55bea52b}{new\+\_\+distance\+\_\+matrix\+\_\+from\+\_\+alignment} (\hyperlink{structalignment__struct}{alignment} align)
\begin{DoxyCompactList}\small\item\em creates and calculates matrix of pairwise distances based on alignment \end{DoxyCompactList}\item 
\mbox{\Hypertarget{alignment_8h_a0d5562e7469a65589c80e20dc561e3ce}\label{alignment_8h_a0d5562e7469a65589c80e20dc561e3ce}} 
void \hyperlink{alignment_8h_a0d5562e7469a65589c80e20dc561e3ce}{store\+\_\+likelihood\+\_\+info\+\_\+at\+\_\+leaf} (double $\ast$$\ast$l, char $\ast$align, int n\+\_\+pat, int n\+\_\+state)
\begin{DoxyCompactList}\small\item\em transform aligned sequence into likelihood for terminal taxa (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) (e.\+g. A -\/$>$ 0001, C-\/$>$ 0010 etc) \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsubsection{Detailed Description}
File handling functions and calculation of distances for sequence data in nexus format. 

Reading of sequence data in nexus format (sequencial or interleaved) and fasta format. For fasta format the sequences don\textquotesingle{}t need to be aligned, but for all formats if the sequences are aligned a data compression is used so that we keep only the distinct site (column) patterns and a mapping between original and compressed site columns. Based on the sequence pairs we can also calculate the matrix of distances between sequences. 