\subsection{File List}
Here is a list of all documented files with brief descriptions\+:\begin{DoxyCompactList}
\item\contentsline{section}{{\bfseries \+\_\+\+\_\+primes\+\_\+unused.\+h} }{\pageref{____primes__unused_8h}}{}
\item\contentsline{section}{{\bfseries \+\_\+header.\+h} }{\pageref{__header_8h}}{}
\item\contentsline{section}{\hyperlink{alignment_8h}{alignment.\+h} \\*File handling functions and calculation of distances for sequence data in nexus format }{\pageref{alignment_8h}}{}
\item\contentsline{section}{{\bfseries argtable3.\+h} }{\pageref{argtable3_8h}}{}
\item\contentsline{section}{\hyperlink{biomcmc_8h}{biomcmc.\+h} \\*Biomcmc library interface to external programs, specific to super\+\_\+sptree repo }{\pageref{biomcmc_8h}}{}
\item\contentsline{section}{\hyperlink{bipartition_8h}{bipartition.\+h} \\*Unary/binary operators on arbitrarily-\/sized bitstrings (strings of zeros and ones) like split bipartitions }{\pageref{bipartition_8h}}{}
\item\contentsline{section}{\hyperlink{char__vector_8c}{char\+\_\+vector.\+c} \\*Vector of strings (species names, leaf names, etc.) }{\pageref{char__vector_8c}}{}
\item\contentsline{section}{\hyperlink{char__vector_8h}{char\+\_\+vector.\+h} \\*List of strings (each string is a vector of chars) }{\pageref{char__vector_8h}}{}
\item\contentsline{section}{\hyperlink{clustering__goptics_8h}{clustering\+\_\+goptics.\+h} \\*O\+P\+T\+I\+CS algorithm based on \href{https://github.com/guineri/GOPTICS}{\tt https\+://github.\+com/guineri/\+G\+O\+P\+T\+I\+CS} }{\pageref{clustering__goptics_8h}}{}
\item\contentsline{section}{\hyperlink{distance__generator_8h}{distance\+\_\+generator.\+h} \\*Distance calculation between generic objects,without generating full matrix beforehand }{\pageref{distance__generator_8h}}{}
\item\contentsline{section}{\hyperlink{distance__matrix_8h}{distance\+\_\+matrix.\+h} \\*Distance matrix, that can be used in alignments and trees, and patristic-\/distance based species distances }{\pageref{distance__matrix_8h}}{}
\item\contentsline{section}{\hyperlink{empirical__frequency_8c}{empirical\+\_\+frequency.\+c} \\*Histogram of vectors, ordered by frequency. Also calculates M\+AP (modal) values }{\pageref{empirical__frequency_8c}}{}
\item\contentsline{section}{\hyperlink{empirical__frequency_8h}{empirical\+\_\+frequency.\+h} \\*Creates a histogram of a vector, ordered by frequency }{\pageref{empirical__frequency_8h}}{}
\item\contentsline{section}{\hyperlink{genetree_8h}{genetree.\+h} \\*Gene tree and species tree structures, for reconciliation etc. This is the high-\/level file with globally exposed functions/structures }{\pageref{genetree_8h}}{}
\item\contentsline{section}{\hyperlink{hashfunctions_8h}{hashfunctions.\+h} \\*Collections of hash functions for 32 and 64 bits, including one-\/liners, murmurhash, and xxhash }{\pageref{hashfunctions_8h}}{}
\item\contentsline{section}{\hyperlink{hashtable_8h}{hashtable.\+h} \\*Double hashing open-\/address hash table using strings as key -- also has distance matrix, that can be used in alignments and trees }{\pageref{hashtable_8h}}{}
\item\contentsline{section}{\hyperlink{hll_8h}{hll.\+h} \\*Hyper\+Log\+Log functions, based on code by Ivan Vitjuk \href{https://github.com/ivitjuk/libhll}{\tt https\+://github.\+com/ivitjuk/libhll} under an I\+SC License }{\pageref{hll_8h}}{}
\item\contentsline{section}{\hyperlink{kmerhash_8h}{kmerhash.\+h} \\*K-\/mer handling of D\+NA sequences, with hash transformation }{\pageref{kmerhash_8h}}{}
\item\contentsline{section}{\hyperlink{lowlevel_8c}{lowlevel.\+c} \\*Lowest level basic functions, that should be available to all other modules }{\pageref{lowlevel_8c}}{}
\item\contentsline{section}{\hyperlink{lowlevel_8h}{lowlevel.\+h} \\*Lowest level header file. Header file for \hyperlink{lowlevel_8c}{lowlevel.\+c} }{\pageref{lowlevel_8h}}{}
\item\contentsline{section}{\hyperlink{newick__space_8h}{newick\+\_\+space.\+h} \\*Reads a list of trees in newick format and creates vector of topologies }{\pageref{newick__space_8h}}{}
\item\contentsline{section}{\hyperlink{nexus__common_8h}{nexus\+\_\+common.\+h} \\*File handling functions for nexus format in general }{\pageref{nexus__common_8h}}{}
\item\contentsline{section}{\hyperlink{parsimony_8h}{parsimony.\+h} \\*Binary and multistate parsimony matrices, together with bipartition extraction for M\+RP }{\pageref{parsimony_8h}}{}
\item\contentsline{section}{\hyperlink{prob__distribution_8h}{prob\+\_\+distribution.\+h} \\*Probability distribution functions and auxiliary mathematical functions from statistical package R }{\pageref{prob__distribution_8h}}{}
\item\contentsline{section}{\hyperlink{prob__distribution__aux_8h}{prob\+\_\+distribution\+\_\+aux.\+h} \\*Auxiliary (low level) functions for prob\+\_\+distribution.\+c }{\pageref{prob__distribution__aux_8h}}{}
\item\contentsline{section}{\hyperlink{quickselect__quantile_8h}{quickselect\+\_\+quantile.\+h} \\*Find k smallest element in vector }{\pageref{quickselect__quantile_8h}}{}
\item\contentsline{section}{\hyperlink{random__number_8h}{random\+\_\+number.\+h} \\*Random number generation, with algorithms from the Gnu Scientific Library (G\+SL) and motivation from the Scalable Parallel Pseudo Random Number Generators Library (S\+P\+R\+NG) }{\pageref{random__number_8h}}{}
\item\contentsline{section}{\hyperlink{random__number__aux_8h}{random\+\_\+number\+\_\+aux.\+h} \\*Variables and structures local to random\+\_\+number.\+c (should be opaque to user) }{\pageref{random__number__aux_8h}}{}
\item\contentsline{section}{{\bfseries random\+\_\+number\+\_\+gen.\+h} }{\pageref{random__number__gen_8h}}{}
\item\contentsline{section}{\hyperlink{read__newick__trees_8h}{read\+\_\+newick\+\_\+trees.\+h} \\*Low-\/level functions for reading newick strings }{\pageref{read__newick__trees_8h}}{}
\item\contentsline{section}{\hyperlink{reconciliation_8h}{reconciliation.\+h} \\*Low-\/level file for gene tree and species tree reconciliation. This file is hidden from user and contains the L\+C\+A-\/based reconciliation distances }{\pageref{reconciliation_8h}}{}
\item\contentsline{section}{\hyperlink{splitset__distances_8h}{splitset\+\_\+distances.\+h} \\*Low-\/level functions that use only the split bipartitions of topologies -- treating them as unrooted usually }{\pageref{splitset__distances_8h}}{}
\item\contentsline{section}{\hyperlink{topology__common_8h}{topology\+\_\+common.\+h} \\*General-\/purpose topology structures created from nexus\+\_\+tree\+\_\+struct (and low-\/level functions) }{\pageref{topology__common_8h}}{}
\item\contentsline{section}{\hyperlink{topology__distance_8h}{topology\+\_\+distance.\+h} \\*Branch length operations on topologies, including patristic distances }{\pageref{topology__distance_8h}}{}
\item\contentsline{section}{\hyperlink{topology__randomise_8h}{topology\+\_\+randomise.\+h} \\*Creation of random topologies and modification of existing ones through branch swapping }{\pageref{topology__randomise_8h}}{}
\item\contentsline{section}{\hyperlink{topology__space_8h}{topology\+\_\+space.\+h} \\*Reads tree files in nexus format and creates a vector of topologies }{\pageref{topology__space_8h}}{}
\item\contentsline{section}{\hyperlink{upgma_8h}{upgma.\+h} \\*U\+P\+G\+MA and bio\+NJ from (onedimensional representation of) distance matrices }{\pageref{upgma_8h}}{}
\end{DoxyCompactList}
